### PR TITLE
EVAKA-4213 Employee attachments

### PR DESCRIPTION
**Pull request by akheron**
#### Summary

Deleting an employee-uploaded attachment doesn't work. Dunno if it should be possible by anyone or just the uploader 🤷

### CODE REVIEW

File: service/src/main/kotlin/fi/espoo/evaka/attachment/AttachmentsController.kt [hunk 1]
```diff
@@ -82,19 +97,23 @@ class AttachmentsController(
...
         db: Database,
         user: AuthenticatedUser,
+        @PathVariable(required = false) incomeStatementId: IncomeStatementId?,
         @RequestPart("file") file: MultipartFile
-    ): ResponseEntity<AttachmentId> {
+    ): AttachmentId {
         Audit.AttachmentsUpload.log(targetId = "nothing")
         user.requireOneOfRoles(UserRole.END_USER)
 
         checkAttachmentCount(db, user)
```
Comment (kimifoud) [inline service/src/main/kotlin/fi/espoo/evaka/attachment/AttachmentsController.kt:113]: We should probably not share the application attachment count limit here.

File: service/src/main/kotlin/fi/espoo/evaka/attachment/AttachmentsController.kt [hunk 2]
```diff
@@ -82,19 +97,23 @@ class AttachmentsController(
...
+        consumes = [MediaType.MULTIPART_FORM_DATA_VALUE]
+    )
     fun uploadAttachmentCitizen(
         db: Database,
         user: AuthenticatedUser,
+        @PathVariable(required = false) incomeStatementId: IncomeStatementId?,
         @RequestPart("file") file: MultipartFile
-    ): ResponseEntity<AttachmentId> {
+    ): AttachmentId {
         Audit.AttachmentsUpload.log(targetId = "nothing")
```
Comment (kimifoud) [inline service/src/main/kotlin/fi/espoo/evaka/attachment/AttachmentsController.kt:110]: `targetId` could be omitted

File: service/src/main/kotlin/fi/espoo/evaka/attachment/AttachmentsController.kt (hunk 1 above)
Comment (kimifoud) [reply service/src/main/kotlin/fi/espoo/evaka/attachment/AttachmentsController.kt:113]: However, we should check access if `incomeStatementId` is provided, right?

File: frontend/src/employee-frontend/components/IncomeStatementPage.tsx [hunk 3]
```diff
@@ -292,30 +348,101 @@ function AccountantInfo({ accountant }: { accountant: Accountant }) {
...
+}) {
+  const { i18n } = useTranslation()
+
+  const handleUpload = React.useCallback(
+    async (
+      file: File,
+      onUploadProgress: (progressEvent: ProgressEvent) => void
+    ) => {
+      return (
+        await saveIncomeStatementAttachment(
```
Comment (kimifoud) [inline frontend/src/employee-frontend/components/IncomeStatementPage.tsx:409]: These handlers might be a little bit more readable with the curly braces removed. 🤷 `=> (await ..).map(..)`

Comment (akheron) [reply frontend/src/employee-frontend/components/IncomeStatementPage.tsx]: Fixed

File: service/src/main/kotlin/fi/espoo/evaka/attachment/AttachmentsController.kt (hunk 2 above)
Comment (akheron) [reply service/src/main/kotlin/fi/espoo/evaka/attachment/AttachmentsController.kt]: Refactored attachment audit logging

File: service/src/main/kotlin/fi/espoo/evaka/attachment/AttachmentsController.kt (hunk 1 above)
Comment (akheron) [reply service/src/main/kotlin/fi/espoo/evaka/attachment/AttachmentsController.kt]: Access check added. Changed count limiting to allow `maxAttachments` per "parent entity" (application, income statement, no parent).

File: frontend/src/employee-frontend/api/attachments.ts [hunk 4]
```diff
@@ -7,31 +7,51 @@ import { UUID } from 'lib-common/types'
...
+    onUploadProgress
+  )
+}
+
+export async function saveIncomeStatementAttachment(
+  incomeStatementId: UUID,
+  file: File,
+  onUploadProgress: (progressEvent: ProgressEvent) => void
+): Promise<Result<UUID>> {
+  return await doSaveAttachment(
```
[the remark this answers is not in the document]
Comment (akheron) [reply frontend/src/employee-frontend/api/attachments.ts:48]: No 😄

**Review by kimifoud — APPROVED** [review]
Solid!

**Pull request merged**